ksql-engine/src/main/java/io/confluent/ksql/util/ExpressionTypeManager.java

```diff
@@ -242,8 +241,38 @@ protected Expression visitFunctionCall(
     return null;
   }
 
-  private Schema resolveArithmeticType(final Schema leftSchema,
+  private static Schema resolveArithmeticType(final Schema leftSchema,
                                        final Schema rightSchema) {
     return SchemaUtil.resolveArithmeticType(leftSchema.type(), rightSchema.type());
   }
+
+  private static void validateComparisonOperandTypes(
+      final ComparisonExpression.Type operator,
+      final Schema.Type leftType,
+      final Schema.Type rightType) {
+    if (SchemaUtil.isNumber(leftType)) {
+      if (!SchemaUtil.isNumber(rightType)) {
+        throw new KsqlException("Invalid comparison operand types. Both sides should be numbers."
+            + " Left type: " + leftType + ", right type: " + rightType);
+      }
+      return;
+    }
+    if (leftType != rightType) {
+      throw new KsqlException("Invalid comparison operand types. Cannot compare incompatible types."
+          + " Left type: " + leftType + ", right type: " + rightType);
+    }
+    if (leftType == Schema.Type.STRING) {
+      return;
+    }
+    if (operator == Type.GREATER_THAN
+        || operator == Type.GREATER_THAN_OR_EQUAL
+        || operator == Type.LESS_THAN
+        || operator == Type.LESS_THAN_OR_EQUAL) {
+      throw new KsqlException("Operator " + operator + " cannot be applied to " + leftType);
+    }
+    if (leftType == Schema.Type.BOOLEAN) {
+      return;
+    }
```

Comment:
Perhaps would be better to put this check in the previous if() block? It would make it more readable if we see that the invalid operators apply to the boolean.

Also, perhaps would better to put what operators apply instead of what do not? I just wonder what if a new operator is added in the future, then that would be allowed to be compare to boolean types because there is no check for that.

Reply:
Perhaps we could create a `Map<Schema.Type, Function<Schema.Type, Boolean>>` populate that with the appropriate functions and then do:
```
 return map.get(leftType).apply(rightType);
```

Reply:
Updated based on @dguy 's suggestion.